File: routa-core/src/main/kotlin/com/phodal/routa/core/pipeline/stages/GateVerificationStage.kt
package com.phodal.routa.core.pipeline.stages

import com.phodal.routa.core.model.AgentRole
import com.phodal.routa.core.model.AgentStatus
import com.phodal.routa.core.model.CompletionReport
import com.phodal.routa.core.model.TaskStatus
import com.phodal.routa.core.model.VerificationVerdict
import com.phodal.routa.core.pipeline.PipelineContext
import com.phodal.routa.core.pipeline.PipelineStage
import com.phodal.routa.core.pipeline.RetryPolicy
import com.phodal.routa.core.pipeline.StageResult
import com.phodal.routa.core.runner.OrchestratorPhase
import com.phodal.routa.core.runner.OrchestratorResult

/**
 * **Stage 4: GATE Verification** — Verifies completed work and decides next steps.
 *
 * This stage:
 * 1. Creates a GATE agent for tasks in REVIEW_REQUIRED status
 * 2. Builds verification context (task definitions, CRAFTER reports, acceptance criteria)
 * 3. Runs the GATE agent to verify all completed work
 * 4. Ensures the verdict is filed (via tool calls or text parsing fallback)
 * 5. Checks the outcome:
 *    - **All approved** → pipeline completes with [StageResult.Done]
 *    - **Some need fixes** → resets tasks and returns [StageResult.RepeatPipeline]
 *    - **No verification needed** → pipeline completes
 *
 * ## Decision flow
 * ```
 * GATE output
 *   → APPROVED for all tasks → Done(Success)
 *   → NOT APPROVED for some  → reset NEEDS_FIX tasks → RepeatPipeline
 * ```
 */
class GateVerificationStage : PipelineStage {

    override val name = "gate-verification"
    override val description = "GATE verifies completed work against acceptance criteria"

    /**
     * Retry policy for GATE verification.
     *
     * Verification LLM calls can fail due to network issues.
     * We retry up to 2 times with exponential backoff.
     */
    override val retryPolicy = RetryPolicy(
        maxAttempts = 2,
        baseDelayMs = 2000,
        backoffMultiplier = 2.0,
    )

    override suspend fun execute(context: PipelineContext): StageResult {
        context.ensureActive() // ← cancellation check before verification
        context.emitPhase(OrchestratorPhase.VerificationStarting(context.waveNumber))

        val gateAgentId = context.system.coordinator.startVerification()
        if (gateAgentId == null) {
            // No tasks need verification — we're done
            context.emitPhase(OrchestratorPhase.Completed)
            return StageResult.Done(buildSuccessResult(context))
        }

        // Build verification context
        val gateContext = buildGateContext(context, gateAgentId)
        val reviewTasks = context.system.context.taskStore.listByStatus(
            context.workspaceId, TaskStatus.REVIEW_REQUIRED
        )
        val taskIdsList = reviewTasks.joinToString(", ") { it.id }
        val gatePrompt = injectAgentIdentity(gateContext, gateAgentId, taskIdsList)

        // Run the GATE agent
        val gateOutput = context.provider.runStreaming(
            AgentRole.GATE, gateAgentId, gatePrompt
        ) { chunk ->
            context.onStreamChunk?.invoke(gateAgentId, chunk)
        }

        // Ensure the verdict is reported
        ensureGateReport(context, gateAgentId, gateOutput)

        // Clean up provider resources
        context.provider.cleanup(gateAgentId)

        context.emitPhase(OrchestratorPhase.VerificationCompleted(gateAgentId, gateOutput))

        // ── Decide next step based on task statuses ─────────────────────

        val allTasks = context.system.context.taskStore.listByWorkspace(context.workspaceId)
        val needsFixTasks = allTasks.filter { it.status == TaskStatus.NEEDS_FIX }

        return when {
            // All tasks completed — success!
            allTasks.isNotEmpty() && allTasks.all { it.status == TaskStatus.COMPLETED } -> {
                context.emitPhase(OrchestratorPhase.Completed)
                StageResult.Done(buildSuccessResult(context))
            }

            // Some tasks need fixes — reset and repeat
            needsFixTasks.isNotEmpty() -> {
                context.emitPhase(OrchestratorPhase.NeedsFix(context.waveNumber))
                resetNeedsFixTasks(context)
                StageResult.RepeatPipeline
            }

            // Other states — continue (might have more pending tasks)
            else -> StageResult.Continue
        }
    }

    // ── Gate report handling ─────────────────────────────────────────────

    /**
     * If the GATE didn't call `report_to_parent` via tool call,
     * parse the verdict from text and file reports on their behalf.
     */
    private suspend fun ensureGateReport(
        context: PipelineContext,
        gateAgentId: String,
        output: String,
    ) {
        val agent = context.system.context.agentStore.get(gateAgentId) ?: return

        // If agent is already COMPLETED, Koog tool calling handled it
        if (agent.status == AgentStatus.COMPLETED) return

        // Parse verdicts from text output
        val reviewTasks = context.system.context.taskStore.listByStatus(
            context.workspaceId, TaskStatus.REVIEW_REQUIRED
        )
        val verdicts = context.reportParser.parseGateVerdicts(gateAgentId, output, reviewTasks)

        // File a report for each task
        for (task in reviewTasks) {
            val verdict = verdicts[task.id]
            val report = CompletionReport(
                agentId = gateAgentId,
                taskId = task.id,
                summary = verdict?.summary ?: "No verdict parsed",
                success = verdict?.verdict == VerificationVerdict.APPROVED,
            )
            context.system.tools.reportToParent(gateAgentId, report)
        }
    }

    // ── Context building ─────────────────────────────────────────────────

    /**
     * Build the verification context for the GATE agent.
     */
    private suspend fun buildGateContext(
        context: PipelineContext,
        gateAgentId: String,
    ): String {
        val roleContext = context.system.coordinator.buildAgentContext(gateAgentId) ?: ""
        val tasks = context.system.context.taskStore.listByStatus(
            context.workspaceId, TaskStatus.REVIEW_REQUIRED
        )

        return buildString {
            appendLine(roleContext)
            appendLine()
            appendLine("---")
            appendLine()
            appendLine("# Tasks to Verify")
            appendLine()

            for (task in tasks) {
                appendLine("## ${task.title}")
                appendLine()
                appendLine("**Task ID:** ${task.id}")
                appendLine()
                appendLine("**Objective:** ${task.objective}")
                appendLine()

                if (task.acceptanceCriteria.isNotEmpty()) {
                    appendLine("**Acceptance Criteria:**")
                    task.acceptanceCriteria.forEach { appendLine("- $it") }
                    appendLine()
                }

                if (task.completionSummary != null) {
                    appendLine("**Crafter Report:** ${task.completionSummary}")
                    appendLine()
                }

                // Include CRAFTER conversation for evidence
                val crafterId = task.assignedTo
                if (crafterId != null) {
                    val conversation = context.system.context.conversationStore.getLastN(crafterId, 5)
                    if (conversation.isNotEmpty()) {
                        appendLine("**Crafter Conversation (last ${conversation.size} messages):**")
                        for (msg in conversation) {
                            appendLine("> [${msg.role}]: ${msg.content.take(500)}")
                        }
                        appendLine()
                    }
                }

                if (task.verificationCommands.isNotEmpty()) {
                    appendLine("**Verification Commands:**")
                    task.verificationCommands.forEach { appendLine("- `$it`") }
                    appendLine()
                }

                appendLine("---")
            }

            appendLine()
            appendLine("Please verify each task against its Acceptance Criteria.")
            appendLine("Output your verdict: ✅ APPROVED or ❌ NOT APPROVED, with evidence.")
        }
    }

    // ── Task reset ───────────────────────────────────────────────────────

    /**
     * Reset NEEDS_FIX tasks back to PENDING for the next wave.
     */
    private suspend fun resetNeedsFixTasks(context: PipelineContext) {
        val tasks = context.system.context.taskStore.listByStatus(
            context.workspaceId, TaskStatus.NEEDS_FIX
        )
        for (task in tasks) {
            context.system.context.taskStore.save(
                task.copy(
                    status = TaskStatus.PENDING,
                    assignedTo = null,
                    updatedAt = java.time.Instant.now().toString(),
                )
            )
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    private fun injectAgentIdentity(prompt: String, agentId: String, taskId: String): String {
        return buildString {
            appendLine("--- Agent Identity ---")
            appendLine("agent_id: $agentId")
            appendLine("task_id: $taskId")
            appendLine("----------------------")
            appendLine()
            append(prompt)
        }
    }

    private suspend fun buildSuccessResult(context: PipelineContext): OrchestratorResult.Success {
        val summary = context.system.coordinator.getTaskSummary()
        return OrchestratorResult.Success(summary)
    }
}

File: routa-js/src/core/pipeline/stages/gate-verification-stage.ts
/**
 * GateVerificationStage - port of routa-core GateVerificationStage.kt
 *
 * Stage 4: GATE Verification - Verifies completed work and decides next steps.
 */

import { AgentRole } from "../../models/agent";
import { TaskStatus } from "../../models/task";
import { PipelineStage } from "../pipeline-stage";
import { PipelineContext, emitPhase } from "../pipeline-context";
import { StageResult } from "../stage-result";

/**
 * Stage 4: GATE Verification
 *
 * This stage:
 * 1. Creates a GATE agent for tasks in REVIEW_REQUIRED status
 * 2. Builds verification context
 * 3. Runs the GATE agent to verify all completed work
 * 4. Checks the outcome:
 *    - All approved → pipeline completes
 *    - Some need fixes → resets tasks and repeats from CRAFTER
 *    - No verification needed → pipeline completes
 */
export class GateVerificationStage implements PipelineStage {
  name = "gate-verification";
  description = "Verifies completed work against acceptance criteria";

  async execute(context: PipelineContext): Promise<StageResult> {
    await emitPhase(context, {
      type: "verification_starting",
      waveNumber: context.waveNumber,
    });

    // Get tasks that need verification
    const reviewTasks = await context.context.taskStore.listByStatus(
      context.workspaceId,
      TaskStatus.REVIEW_REQUIRED
    );

    if (reviewTasks.length === 0) {
      // No tasks need verification — we're done
      await emitPhase(context, { type: "completed" });
      const allTasks = await context.context.taskStore.listByWorkspace(
        context.workspaceId
      );
      return StageResult.Done({
        type: "success",
        taskSummaries: allTasks.map((t) => t.title),
        planOutput: context.planOutput,
      });
    }

    // Create GATE agent
    const result = await context.context.agentTools.createAgent({
      name: `gate-wave-${context.waveNumber}`,
      role: AgentRole.GATE,
      workspaceId: context.workspaceId,
      parentId: context.routaAgentId,
    });

    if (!result.success || !result.data) {
      return StageResult.Failed("Failed to create GATE agent");
    }

    const gateAgentId = (result.data as { agentId: string }).agentId;
    context.gateAgentId = gateAgentId;

    // Build verification context
    const gatePrompt = await this.buildGatePrompt(context, reviewTasks);

    // Execute GATE
    const gateOutput = await context.provider.runStreaming(
      AgentRole.GATE,
      gateAgentId,
      gatePrompt,
      (chunk) => {
        if (context.onStreamChunk) {
          context.onStreamChunk(gateAgentId, chunk);
        }
      }
    );

    // Parse GATE verdict (simplified - just check for "APPROVED" keyword)
    const allApproved = gateOutput.toUpperCase().includes("APPROVED");

    if (allApproved) {
      // Mark all tasks as COMPLETED
      for (const task of reviewTasks) {
        await context.context.taskStore.save({
          ...task,
          status: TaskStatus.COMPLETED,
        });
      }

      await emitPhase(context, { type: "completed" });
      return StageResult.Done({
        type: "success",
        taskSummaries: reviewTasks.map((t) => t.title),
        planOutput: context.planOutput,
      });
    } else {
      // Some tasks need fixes — reset and repeat
      for (const task of reviewTasks) {
        await context.context.taskStore.save({
          ...task,
          status: TaskStatus.NEEDS_FIX,
        });
      }

      await emitPhase(context, {
        type: "needs_fix",
        waveNumber: context.waveNumber,
      });
      return StageResult.RepeatPipeline("crafter-execution");
    }
  }

  private async buildGatePrompt(
    context: PipelineContext,
    reviewTasks: any[]
  ): Promise<string> {
    const taskSummaries = reviewTasks
      .map(
        (task) => `
# ${task.title}

## Objective
${task.objective}

## Acceptance Criteria
${task.acceptanceCriteria.map((c: string) => `- ${c}`).join("\n")}

## Verification Commands
${task.verificationCommands.map((c: string) => `- ${c}`).join("\n")}
`.trim()
      )
      .join("\n\n---\n\n");

    return `
You are a GATE agent. Your role is to verify that the following tasks have been completed correctly.

For each task, check:
1. All acceptance criteria are met
2. Verification commands pass
3. Code quality is acceptable

${taskSummaries}

## Your Response

If all tasks are approved, respond with "✅ APPROVED".
If any tasks need fixes, respond with "❌ NEEDS FIX" and explain what needs to be fixed.
`.trim();
  }
}

